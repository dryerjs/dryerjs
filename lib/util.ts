import { plural, singular } from './js/pluralize';

export { plural, singular };

export const isNil = (value: any): value is null | undefined => value === null || value === undefined;

export const isNotNil = (value: any) => !isNil(value);

export const isObject = (value: any): value is object => typeof value === 'object';

export const isArray = (value: any) => Array.isArray(value);

export const isNotNullObject = (value: any): value is object => typeof value === 'object' && value !== null;

export const isFunction = (value: any) => typeof value === 'function';

export const isUndefined = (value: any): value is undefined => typeof value === 'undefined';

export const isTruthy = (value: any) => !!value;

export const isString = (value: any): value is string => typeof value === 'string';

export const defaultTo = <T>(value: T | null | undefined, defaultValue: T): T => {
  const shouldUseDefaultValue = isNil(value) || (typeof value === 'number' && isNaN(value));
  return shouldUseDefaultValue ? defaultValue : (value as T);
};

export const deepOmit = (object: any, omitKeys: string[]) => {
  if (typeof object !== 'object' || object === null) {
    return object;
  }

  const result = Array.isArray(object) ? [] : {};

  for (const key in object) {
    if (object.hasOwnProperty(key) && !omitKeys.includes(key)) {
      if (typeof object[key] === 'object' && object[key] !== null) {
        // Recursively omit keys from nested objects
        result[key] = deepOmit(object[key], omitKeys);
      } else {
        result[key] = object[key];
      }
    }
  }

  return result;
};

export const toPascalCase = (str: string) => str.replace(str[0], str[0].toUpperCase());

export const toCamelCase = (str: string) => str.replace(str[0], str[0].toLowerCase());

export const last = <T>(array: T[]) => array[array.length - 1];

export function memoize<T extends Function>(func: T, resolver?: (...args: any[]) => any): T {
  /* istanbul ignore if */
  if (typeof func !== 'function' || (resolver && typeof resolver !== 'function')) {
    throw new TypeError('FUNC_ERROR_TEXT');
  }

  const memoized = function (this: any, ...args: any[]) {
    const key = resolver ? resolver.apply(this, args) : String(args[0]);
    const cache = memoized.cache;
    if (cache.has(key)) return cache.get(key);
    const result = func.apply(this, args);
    cache.set(key, result);
    return result;
  };

  memoized.cache = new Map();
  return memoized as any;
}
