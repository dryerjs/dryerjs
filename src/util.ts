export { plural, singular } from 'pluralize';

export const isNil = (value: any): value is null | undefined => value === null || value === undefined;

export const isNotNil = (value: any) => !isNil(value);

export const isObject = (value: any): value is object => typeof value === 'object';

export const isNotNullObject = (value: any): value is object => typeof value === 'object' && value !== null;

export const isFunction = (value: any) => typeof value === 'function';

export const isUndefined = (value: any): value is undefined => typeof value === 'undefined';

export const isTruthy = (value: any) => !!value;

export const isString = (value: any): value is string => typeof value === 'string';

export const defaultTo = <T>(value: T | null | undefined, defaultValue: T): T => {
    return value != null && !isNaN(value as any) ? value : defaultValue;
};

export const isEnumOfNumbers = (enumObject: any) => {
    return Object.values(enumObject).some((value: string) => '0123456789'.includes(value));
};

export const deepOmit = (object: any, omitKeys: string[]) => {
    if (typeof object !== 'object' || object === null) {
        return object;
    }

    const result = Array.isArray(object) ? [] : {};

    for (const key in object) {
        if (object.hasOwnProperty(key) && !omitKeys.includes(key)) {
            if (typeof object[key] === 'object' && object[key] !== null) {
                // Recursively omit keys from nested objects
                result[key] = deepOmit(object[key], omitKeys);
            } else {
                result[key] = object[key];
            }
        }
    }

    return result;
};

export const toPascalCase = (str: string) => str.replace(str[0], str[0].toUpperCase());

export const toCamelCase = (str: string) => str.replace(str[0], str[0].toLowerCase());

export const last = <T>(array: T[]) => array[array.length - 1];
