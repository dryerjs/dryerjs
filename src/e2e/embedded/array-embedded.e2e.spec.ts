import * as graphql from 'graphql';
import {
    DefaultOnInput,
    EmbeddedProperty,
    ExcludeOnCreate,
    NullableOnOutput,
    Property,
    RequiredOnCreate,
    TransformOnInput,
    Validate,
} from 'dryerjs';
import * as util from '../../util';
import { DryerTest } from '../dryer-test';

class Book {
    @ExcludeOnCreate()
    @Property()
    id: string;

    @Property()
    @TransformOnInput((name: string) => name.trim())
    name: string;

    @Property()
    @NullableOnOutput()
    @Validate((publishedYear: number) => {
        if (publishedYear >= 0) return;
        throw new graphql.GraphQLError(`Invalid publishedYear ${publishedYear}`, {
            extensions: {
                code: 'BAD_REQUEST',
                http: { status: 400 },
            },
        });
    })
    @DefaultOnInput(() => 0)
    publishedYear: number;
}

class Author {
    @ExcludeOnCreate()
    @Property()
    id: string;

    @EmbeddedProperty({ type: Book })
    @NullableOnOutput()
    books: Book[];

    @Property()
    @RequiredOnCreate()
    name: string;
}

export const dryer = DryerTest.init({
    modelDefinitions: [Author],
});

describe('Object embedded feature works', () => {
    beforeAll(async () => {
        await dryer.start();
    });

    describe('User CRUD works', () => {
        let allAuthors: Author[];

        beforeAll(async () => {
            const authorInputs = [
                {
                    name: 'Author 1',
                    books: [
                        {
                            name: 'Book 1',
                            publishedYear: 2000,
                        },
                        {
                            name: 'Book 2',
                        },
                    ],
                },
                {
                    name: 'Author 2',
                },
            ];

            for (const authorInput of authorInputs) {
                await dryer.makeSuccessRequest({
                    query: `
                        mutation CreateAuthor($input: CreateAuthorInput!) {
                            createAuthor(input: $input) {
                                id
                            }
                        }
                    `,
                    variables: { input: authorInput },
                });
            }

            const allAuthorsResponse = await dryer.makeSuccessRequest({
                query: `
                    query Authors {
                        allAuthors {
                            id
                            name
                            books {
                                id
                                name
                                publishedYear
                            }
                        }
                    }
                `,
            });

            allAuthors = allAuthorsResponse.allAuthors;
        });

        it('should show authors with books', async () => {
            expect(util.deepOmit(allAuthors, ['id'])).toMatchSnapshot();
        });

        it('should be able to update books', async () => {
            const firstAuthor = allAuthors[0];
            const firstBook = firstAuthor.books[0];
            const firstUpdateResponse = await dryer.makeSuccessRequest({
                query: `
                    mutation UpdateAuthor($input: UpdateAuthorInput!) {
                        updateAuthor(input: $input) {
                            books {
                                id
                                name
                                publishedYear
                            }
                        }
                    }
                `,
                variables: {
                    input: {
                        id: firstAuthor.id,
                        books: [
                            {
                                ...firstBook,
                                publishedYear: 1999,
                            },
                        ],
                    },
                },
            });
            expect(firstUpdateResponse.updateAuthor.books[0]).toEqual({
                id: firstBook.id,
                name: 'Book 1',
                publishedYear: 1999,
            });
        });

        describe('embedded apis work', () => {
            let author: Author;

            beforeAll(async () => {
                const authorResponse = await dryer.makeSuccessRequest({
                    query: `
                        mutation CreateAuthor($input: CreateAuthorInput!) {
                            createAuthor(input: $input) {
                                id
                                name
                                books {
                                    id
                                    name
                                    publishedYear
                                }
                            }
                        }
                    `,
                    variables: {
                        input: {
                            name: 'Author 3',
                            books: [
                                {
                                    name: 'Book 1',
                                    publishedYear: 2000,
                                },
                                {
                                    name: 'Book 2',
                                },
                            ],
                        },
                    },
                });
                author = authorResponse.createAuthor;
            });

            it('getOne works', async () => {
                const { authorBook } = await dryer.makeSuccessRequest({
                    query: `
                        query AuthorBook($id: String!, $authorId: String!) {
                            authorBook(id: $id, authorId: $authorId) {
                                id
                                name
                                publishedYear
                            }
                        }
                    `,
                    variables: { authorId: author.id, id: author.books[0].id },
                });
                expect(authorBook).toEqual({
                    id: author.books[0].id,
                    name: 'Book 1',
                    publishedYear: 2000,
                });
            });

            it('create works', async () => {
                const response = await dryer.makeSuccessRequest({
                    query: `
                        mutation CreateAuthorBook($input: CreateBookInput!, $authorId: String!) {
                            createAuthorBook(input: $input, authorId: $authorId) {
                                id
                                name
                                publishedYear
                            }
                        }
                    `,
                    variables: {
                        input: {
                            name: 'Book 3',
                            publishedYear: 2003,
                        },
                        authorId: author.id,
                    },
                });
                expect(response.createAuthorBook).toEqual({
                    id: expect.any(String),
                    name: 'Book 3',
                    publishedYear: 2003,
                });
            });

            it('update works', async () => {
                const response = await dryer.makeSuccessRequest({
                    query: `
                        mutation UpdateAuthorBook($input: UpdateBookInput!, $authorId: String!) {
                            updateAuthorBook(input: $input, authorId: $authorId) {
                                id
                                name
                            }
                        }
                    `,
                    variables: {
                        input: {
                            name: 'Book 1 updated',
                            id: author.books[0].id,
                        },
                        authorId: author.id,
                    },
                });
                expect(response.updateAuthorBook).toEqual({
                    id: author.books[0].id,
                    name: 'Book 1 updated',
                });
            });

            it('delete works', async () => {
                const book2Id = author.books[1].id;
                const response = await dryer.makeSuccessRequest({
                    query: `
                        mutation DeleteAuthorBook($authorId: String!, $id: String!) {
                            deleteAuthorBook(authorId: $authorId, id: $id) {
                                id
                                deleted
                            }
                        }
                    `,
                    variables: {
                        authorId: author.id,
                        id: book2Id,
                    },
                });
                expect(response.deleteAuthorBook).toEqual({
                    id: book2Id,
                    deleted: true,
                });
            });

            it('get all works', async () => {
                const { allAuthorBooks } = await dryer.makeSuccessRequest({
                    query: `
                        query AuthorBook($authorId: String!) {
                            allAuthorBooks(authorId: $authorId) {
                                name
                                publishedYear
                            }
                        }
                    `,
                    variables: { authorId: author.id, id: author.books[0].id },
                });
                expect(allAuthorBooks).toEqual([
                    { name: 'Book 1 updated', publishedYear: 2000 },
                    { name: 'Book 3', publishedYear: 2003 },
                ]);
            });

            const notFoundId = '000000000000000000000000';

            it('getOne on not found', async () => {
                await dryer.makeFailRequest({
                    query: `
                        query AuthorBook($id: String!, $authorId: String!) {
                            authorBook(id: $id, authorId: $authorId) {
                                id
                                name
                                publishedYear
                            }
                        }
                    `,
                    variables: { authorId: author.id, id: notFoundId },
                    errorMessageMustContains: 'No book found with id',
                });
            });

            it('delete on not found', async () => {
                await dryer.makeFailRequest({
                    query: `
                        mutation DeleteAuthorBook($authorId: String!, $id: String!) {
                            deleteAuthorBook(authorId: $authorId, id: $id) {
                                id
                                deleted
                            }
                        }
                    `,
                    variables: { authorId: author.id, id: notFoundId },
                    errorMessageMustContains: 'No book found with id',
                });
            });

            it('update on not found', async () => {
                await dryer.makeFailRequest({
                    query: `
                        mutation UpdateAuthorBook($input: UpdateBookInput!, $authorId: String!) {
                            updateAuthorBook(input: $input, authorId: $authorId) {
                                id
                                name
                            }
                        }
                    `,
                    variables: {
                        input: {
                            name: 'does not matter',
                            id: notFoundId,
                        },
                        authorId: author.id,
                    },
                    errorMessageMustContains: 'No book found with id',
                });
            });
        });
    });

    afterAll(async () => {
        await dryer.stop();
    });
});
