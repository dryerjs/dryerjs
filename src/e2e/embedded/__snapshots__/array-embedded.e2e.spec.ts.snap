// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object embedded feature works User CRUD works should show authors with books 1`] = `
[
  {
    "books": [
      {
        "name": "Book 1",
        "publishedYear": 2000,
      },
      {
        "name": "Book 2",
        "publishedYear": 0,
      },
    ],
    "name": "Author 1",
  },
  {
    "books": [],
    "name": "Author 2",
  },
]
`;
